Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -491,20 +411,72 @@ extension SwiftLanguageServer {
     req[keys.request] = self.requests.editor_close
     req[keys.name] = uri.pseudoPath
 
-    // Clear settings that should not be cached for closed documents.
-    self.currentDiagnostics[uri] = nil
-
     _ = try? self.sourcekitd.sendSync(req)
 
     await semanticTokensManager.discardSemanticTokens(for: note.textDocument.uri)
   }
 
+  /// Cancels any in-flight tasks to send a `PublishedDiagnosticsNotification` after edits.
+  ///
+  /// - Important: Cancelling in-flight publish diagnostics tasks is required before any edit. See the comment inside
+  ///   `documentDiagnostic` for more details.
+  private func cancelInFlightPublishDiagnosticsTask(for document: DocumentURI) {
+    if let inFlightTask = inFlightPublishDiagnosticsTasks[document] {
+      inFlightTask.cancel()
+    }
+  }
+
+  /// If the client doesn't support pull diagnostics, compute diagnostics for the latest version of the given document
+  /// and send a `PublishDiagnosticsNotification` to the client for it.
+  private func publishDiagnosticsIfNeeded(for document: DocumentURI) {
+    guard enablePublishDiagnostics else {
+      return
+    }
+    guard diagnosticsEnabled(for: document) else {
+      logger.debug("Ignoring diagnostics for blacklisted file \(document.pseudoPath)")
+      return
+    }
+    cancelInFlightPublishDiagnosticsTask(for: document)
+    inFlightPublishDiagnosticsTasks[document] = Task(priority: .medium) { [weak self] in
+      guard let self, let sourceKitServer = await self.sourceKitServer else {
+        logger.fault("Cannot produce PublishDiagnosticsNotification because sourceKitServer was deallocated")
+        return
+      }
+      do {
+        // Sleep for a little bit until triggering the diagnostic generation. This effectively de-bounces diagnostic
+        // generation since any later edit will cancel the previous in-flight task, which will thus never go on to send
+        // the `DocumentDiagnosticsRequest`.
+        try await Task.sleep(for: sourceKitServer.options.swiftPublishDiagnosticsDebounceDuration)
+      } catch {
+        return
+      }
+      do {
+        // Run the diagnostic report request through `SourceKitServer` so that it gets handled on `SourceKitServer`'s
+        // `messageHandlingQueue`. This ensures that the diagnostics request will not run concurrently with any document
+        // edits.
+        let diagnosticReport = try await self.fullDocumentDiagnosticReport(DocumentDiagnosticsRequest(textDocument: TextDocumentIdentifier(document)))
```

Comment:
Is this comment outdated? AFAICT we don't go through the `messageHandlingQueue`

Reply:
Thanks for catching it.